### ISSUE TITLE
Request to discuss SVI

### ISSUE THREAD

**Issue by LeoLiu1998** [wohlert/semi-supervised-pytorch#15]
Hi Wohlert, I believe the [code](https://github.com/wohlert/semi-supervised-pytorch/blob/7cde4959468d271552febdcfed5e1cfae9857613/semi-supervised/inference/variational.py#L97) connects to part of Eq. 20 in the paper -- <a href="https://www.codecogs.com/eqnedit.php?latex=E_{q_\phi(z|x)}&space;[\log&space;p_{\theta}(x|z)]" target="_blank"><img src="https://latex.codecogs.com/gif.latex?E_{q_\phi(z|x)}&space;[\log&space;p_{\theta}(x|z)]" title="E_{q_\phi(z|x)} [\log p_{\theta}(x|z)]" /></a>, but I am not entirely sure why that's how we implement the formular, especially why we want to add the prior. Could you explain it a bit? Which paper are you writing code for?

**Comment by wohlert** [comment]
The SVI class is only related to semi-supervised learning such as the one presented in [Kingma 2014](https://arxiv.org/abs/1406.5298).

I don't believe this is the appropriate forum to discuss paper details. If your posts are not directly related to bugs/errors in the code or feature requests, then please refrain from creating issues.

Thanks

**Comment by LeoLiu1998** [comment]
I am sorry for such confusion.

I meant to ask why you have ['prior'](https://github.com/wohlert/semi-supervised-pytorch/blob/7cde4959468d271552febdcfed5e1cfae9857613/semi-supervised/inference/variational.py#L97) when calculating <a href="https://www.codecogs.com/eqnedit.php?latex=E_{q_\phi(z|x)}&space;[\log&space;p_{\theta}(x|z)]" target="_blank"><img src="https://latex.codecogs.com/gif.latex?E_{q_\phi(z|x)}&space;[\log&space;p_{\theta}(x|z)]" title="E_{q_\phi(z|x)} [\log p_{\theta}(x|z)]" /></a>, I thought calculating with cross entropy is enough. I check other people's [code](https://github.com/wiseodd/generative-models/blob/b930d5fa9e2f69adfd4ea8ec759f38f6ce6da4c2/VAE/vanilla_vae/vae_pytorch.py#L84) to implement the loss function, the didn't include a prior.

**Comment by wohlert** [comment]
Because SVI is for the VAE but for the semi-supervised VAE. The prior is over the label.

**Comment by LeoLiu1998** [comment]
Gotcha, could you also tell me which paper the code intends to work for? No worry if the code is of general purpose.

**Comment by wohlert** [comment]
It's the article linked in my first comment.

**Comment by lithostark** [comment]
> 
> 
> I am sorry for such confusion.
> 
> I meant to ask why you have ['prior'](https://github.com/wohlert/semi-supervised-pytorch/blob/7cde4959468d271552febdcfed5e1cfae9857613/semi-supervised/inference/variational.py#L97) when calculating [![](https://camo.githubusercontent.com/33d92a5ed9dea9166b0a9f98b966c59a25060e9f/68747470733a2f2f6c617465782e636f6465636f67732e636f6d2f6769662e6c617465783f455f7b715f5c706869287a7c78297d2673706163653b5b5c6c6f672673706163653b705f7b5c74686574617d28787c7a295d)](https://www.codecogs.com/eqnedit.php?latex=E_{q_\phi(z|x)}&space;[\log&space;p_{\theta}(x|z)]), I thought calculating with cross entropy is enough. I check other people's [code](https://github.com/wiseodd/generative-models/blob/b930d5fa9e2f69adfd4ea8ec759f38f6ce6da4c2/VAE/vanilla_vae/vae_pytorch.py#L84) to implement the loss function, the didn't include a prior.

Hi, I'm having trouble understanding why calculating 
![image](https://user-images.githubusercontent.com/37433550/56571199-07905780-65ef-11e9-9b62-3e265a3a74f8.png) with cross entropy. The expectation is calculated over q(z|x), a distribution over z, but p(x|z) is a distribution over x. Could you give me some hint please?

**Comment by wohlert** [comment]
p(x|z) is no more a distribution over x than it is over z. In this case the x in the expectation is observed, so taking the expectation with respect to this variable is not helpful for our model. It might help you to consider how we would write this in terms of a Gaussian likelihood.

<img src="https://latex.codecogs.com/gif.latex?\mathbb{E}_{q_{\phi}(z|x)}[p_{\theta}(x|z)]&space;\approx&space;\frac{1}{n}\sum_{i=1}^{n}||x&space;-&space;g_{\theta}(z_i)||_2^2" title="\mathbb{E}_{q_{\phi}(z|x)}[p_{\theta}(x|z)] \approx \frac{1}{n}\sum_{i=1}^{n}||x - g_{\theta}(z_i)||_2^2" />

Where g is the decoder and we are taking a Monte Carlo estimate of the expectation.

**Comment by lithostark** [comment]
Thank you for your explanation! You are really a nice person~ So the cross entropy is actually between the empirical distribution over x and model distribution over x. The q(z|x) only deals with MC estimation. Is that correct?

**Comment by wohlert** [comment]
Yes, this is also what we call the likelihood of the data under the model.
